### PR TITLE
feat(core): Augment data instead of copying it

### PR DESCRIPTION
**Pull request by janober**
Currently, Code & Function-Nodes copy a lot of data just to make sure that not by accident anything that should not change gets changed. This PR will enable the same but just in a cheaper way.

### CODE REVIEW

**Review by flipswitchingmonkey — CHANGES_REQUESTED** [review]
I think this approach works, but it feels like we are re-inventing a number of things that we could get out of the box by using prototype chaining instead. As an example:

```ts
interface SomeObject {
  [key:string]: any;
  title: string;
  version: string;
  data: number;
}

const parent: SomeObject = {
  title: "Parent",
  version: "1",
  data: 1,
}
Object.freeze(parent);

const child1: SomeObject = Object.create(parent);
child1.title = "Child1";
child1.version = "1.1";
parent.data = 5;  // does not change!

const child2: SomeObject = Object.create(parent);
child2.title = "Child2";
child2.version = "1.2";

Object.freeze(child1);
child1.title = "Changed"; // does not change!

const grandChild: SomeObject = Object.create(child1);
grandChild.title = "GrandChild";
grandChild.version = "1.1.1";

function print(o: SomeObject) {
  console.log(`${o.title} version: ${o.version} data: ${o.data} parent: ${o.__proto__.title}`);
}

print(parent); // 'Parent version: 1 data: 1 parent: undefined'
print(child1); // 'Child1 version: 1.1 data: 1 parent: Parent'
print(child2); // 'Child2 version: 1.2 data: 1 parent: Parent'
print(grandChild); // 'GrandChild version: 1.1.1 data: 1 parent: Child1'
```

Edit: I can see how this does not solve the recursive freezing and deleting of properties problem, so the Proxy solution is probably the right way 👍

**Review by netroy — COMMENTED** [review]
We might be able to replace reflect-metadata calls with prototype chaining for possibly better performance. but that can be done in another PR.

**Review by flipswitchingmonkey — DISMISSED** [review]

File: packages/workflow/src/WorkflowDataProxy.ts [hunk 1]
```diff
@@ -96,9 +97,13 @@ export class WorkflowDataProxy {
...
-			? deepCopy(runExecutionData)
+			? runExecutionData !== null
+				? augmentObject(runExecutionData)
+				: null
 			: runExecutionData;
 
+		// This is slower with augmentArray than deepCopy because literally all
+		// data gets always accessed, so leave for now
 		this.connectionInputData = isScriptingNode(activeNodeName, workflow)
 			? deepCopy(connectionInputData)
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/workflow/src/WorkflowDataProxy.ts:108]: after testing with a bunch of different JSON sizes, and multiple code-nodes in a workflow, I'm certain that the pros of using `augmentArray` here significantly out-weigh any cons.

**Review by netroy — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@0.221.2`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.221.2)

**Comment by janober** [comment]
Got released with [`n8n@0.221.2`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.221.2)

**Pull request merged**